### PR TITLE
Cookie route

### PR DESCRIPTION
**Pull request by ericdrosado**

### CODE REVIEW

File: Sources/Server/Parser.swift [hunk 1]
```diff
@@ -1,32 +1,68 @@
...
+            return Request(method: String(requestComponents[0]), path: parsedPathWithQueries[0], queries: queries) 
+        } else if (headers.keys.contains("Cookie")) {
+            let cookie = parseCookie(headers: headers)
+            return Request(method: String(requestComponents[0]), path: String(requestComponents[1]), cookie: cookie["type"]!) 
+        } else {
+            return Request(method: String(requestComponents[0]), path: String(requestComponents[1]))
+        }
+    }
+
+    private func parseRequest(request: String) -> (Array<Substring>, Array<String>) {
```
Comment (cdemyanovich) [inline Sources/Server/Parser.swift]: Although Swift seems to allow functions with the same name and arguments but different return types, I think I'd prefer each function had a unique name to better be able to distinguish them. Thoughts, @tommcgee?

Also, is `Array<Substring>` legit?

File: Sources/Server/Parser.swift [hunk 2]
```diff
@@ -1,32 +1,68 @@
...
+        var requestAndHeaders = request.components(separatedBy: "\r\n")
+        let requestComponents = requestAndHeaders[0].split(separator: " ")
+        let requestHeaders = parseNonHeaderComponents(requestWithHeaders: requestAndHeaders)
+        return (requestComponents, requestHeaders)
+    }
+
+    private func parseNonHeaderComponents(requestWithHeaders: Array<String>) -> (Array<String>) {
+        var headers = requestWithHeaders
+        headers.removeFirst(1)
+        headers.removeLast(2)
```
Comment (cdemyanovich) [inline Sources/Server/Parser.swift]: As I mentioned before, find a way to clarify why `removeFirst` and `removeLast` are needed here. Ideally, you would NOT just use a comment. :-)

File: Sources/Server/Parser.swift [hunk 3]
```diff
@@ -1,32 +1,68 @@
...
+        headers.removeFirst(1)
+        headers.removeLast(2) 
+        return headers 
+    }
+
+    private func parsePath(path: String) -> Array<String> {
         let parsedPathWithQueries = path.split(separator: "?").map(String.init) 
         return parsedPathWithQueries
-   }
+    }
```
Comment (cdemyanovich) [inline Sources/Server/Parser.swift]: You could just return the value instead of using a variable.

File: Sources/Server/Parser.swift (hunk 2 above)
Comment (ericdrosado) [reply Sources/Server/Parser.swift]: I thought "parseNonHeaderComponents" would be a good enough explanation... back to the drawing board.

File: Sources/Server/Parser.swift (hunk 1 above)
Comment (ericdrosado) [reply Sources/Server/Parser.swift]: Whoops! My mistake. Yes, `Array<Substring>` is a thing. I could convert it to an `Array<String>`. I'll go ahead and do that.

File: Sources/Server/Parser.swift (hunk 2 above)
Comment (cdemyanovich) [reply Sources/Server/Parser.swift]: What are non-header components, though?

Comment (ericdrosado) [reply Sources/Server/Parser.swift]: Perhaps use two methods where `headers.removeFirst(1)`, which removes the HTTP request line, could be `removeHTTPRequest()` and `headers.removeLast(2)`, which removes the two new lines, could be `removeNewLine()`

Comment (cdemyanovich) [reply Sources/Server/Parser.swift]: Given this example raw request from one of my projects
```
GET /api/rest/v5/agreements/abc123 HTTP/1.1
Access-Token: foo
Host: api.na1.echosign.com
Connection: close
User-Agent: Paw/3.1.5 (Macintosh; OS X/10.13.3) GCDHTTPRequest


```
what do you mean by "the HTTP request line"?

Comment (ericdrosado) [reply Sources/Server/Parser.swift]: "HTTP request line" would equal first line: `GET /api/rest/v5/agreements/abc123 HTTP/1.1`. In my mind, which may be completely wrong, I labeled the first line as the actual "request" and the following lines as the "headers"

**Review by cdemyanovich — APPROVED** [review]
Since you've addressed my earlier feedback, I'll approve. But, I know @tommcgee is planning to review, too.

File: Tests/Test/ResponseTest.swift [hunk 4]
```diff
@@ -7,59 +7,60 @@ class ResponseTest: XCTestCase {
     let response = Response(parser: parser)
     static let body200 = "<!DOCTYPE html><html><body><h1>Hello World</h1></body></html>"
     static let body404 = "<!DOCTYPE html><html><body><h1>404 Page Not Found</h1></body></html>"
+    let requestHeader = "\r\nCache-Control: no-cache\r\nConnection: keep-alive\r\n\r\n"
```
Comment (tommcgee) [inline Tests/Test/ResponseTest.swift:10]: Think about abstracting `\r\n` to a constant. With this very specific & important pattern 'magic stringed' all over the place you're leaving yourself very vulnerable to typos.

**Review by tommcgee — APPROVED** [review]
Besides some minor comments, looks good to me.

**Pull request merged**